lighthouse-core/gather/driver/wait-for-condition.js

```diff
@@ -0,0 +1,420 @@
+    throw new Error('waitForNetworkIdle.cancel() called before it was defined');
+  };
+
+  /** @type {Promise<void>} */
+  const promise = new Promise((resolve, reject) => {
+    const onIdle = () => {
+      // eslint-disable-next-line no-use-before-define
+      networkMonitor.once('network-2-busy', onBusy);
+      idleTimeout = setTimeout(() => {
+        cancel();
+        resolve();
+      }, networkQuietThresholdMs);
+    };
+
+    const onBusy = () => {
+      networkMonitor.once('network-2-idle', onIdle);
+      idleTimeout && clearTimeout(idleTimeout);
+    };
+
+    const domContentLoadedListener = () => {
+      hasDCLFired = true;
+      if (networkMonitor.is2Idle()) {
+        onIdle();
+      } else {
+        onBusy();
+      }
+    };
+
+    // We frequently need to debug why LH is still waiting for the page.
+    // This listener is added to all network events to verbosely log what URLs we're waiting on.
+    const logStatus = () => {
+      if (!hasDCLFired) {
+        log.verbose('waitFor', 'Waiting on DomContentLoaded');
+        return;
+      }
+
+      const inflightRecords = networkMonitor.getInflightRecords();
+      // If there are more than 20 inflight requests, load is still in full swing.
+      // Wait until it calms down a bit to be a little less spammy.
+      if (inflightRecords.length < 20) {
+        for (const record of inflightRecords) {
+          log.verbose('waitFor', `Waiting on ${record.url.slice(0, 120)} to finish`);
+        }
+      }
+    };
+
+    networkMonitor.on('requeststarted', logStatus);
+    networkMonitor.on('requestloaded', logStatus);
+    networkMonitor.on('network-2-busy', logStatus);
+
+    session.once('Page.domContentEventFired', domContentLoadedListener);
+    let canceled = false;
+    cancel = () => {
+      if (canceled) return;
+      canceled = true;
+      idleTimeout && clearTimeout(idleTimeout);
+      session.off('Page.domContentEventFired', domContentLoadedListener);
+      networkMonitor.removeListener('network-2-busy', onBusy);
+      networkMonitor.removeListener('network-2-idle', onIdle);
+      networkMonitor.removeListener('requeststarted', logStatus);
+      networkMonitor.removeListener('requestloaded', logStatus);
+      networkMonitor.removeListener('network-2-busy', logStatus);
+    };
+  });
+
+  return {
+    promise,
+    cancel,
+  };
+}
+
+/**
+ * Resolves when there have been no long tasks for at least waitForCPUQuiet ms.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {number} waitForCPUQuiet
+ * @return {CancellableWait}
+ */
+function waitForCPUIdle(session, waitForCPUQuiet) {
+  if (!waitForCPUQuiet) {
+    return {
+      promise: Promise.resolve(),
+      cancel: () => undefined,
+    };
+  }
+
+  /** @type {NodeJS.Timer|undefined} */
+  let lastTimeout;
+  let canceled = false;
+
+  const checkForQuietExpression = `(${pageFunctions.checkTimeSinceLastLongTaskString})()`;
+  /**
+   * @param {ExecutionContext} executionContext
```

Comment:
directly using `executionContext` is an FR-compat change and functionally equivalent to using driver when isolation is disabled